Disable parser cache by default

arporter opened the pull request:
This is a very small change that adds a new argument to `api.parse()` which allows the use of the fparser cache to be controlled. By default the cache is now wiped at the beginning of each call to this routine.

MatthewHambley commented:
Please briefly explain the reason for wanting this? What problem does not clearing the cache cause?

rupertford commented:
Hi @MatthewHambley, the issue that prompted this change was reported in https://github.com/stfc/fparser/issues/74.

In summary, we generate code using a wrapper library around fparser called f2pygen. This library reads in "vanilla" snippets of code and then modifies them. In Python3 the caching mechanism of fparser was keeping previously modified snippets of code rather than starting with vanilla versions so we were getting incorrect output.

arporter commented:
It's also worth noting that PSyclone already does this wiping of the cache manually before parsing of Algorithm and Kernel code.

arporter commented on src/fparser/tests/test_api.py:
```diff
@@ -66,24 +67,42 @@
-    end
     '''
-
+    Tests for parsing the `dimension` attribute of variable declarations
```
Done

arporter commented on src/fparser/tests/test_api.py:
```diff
@@ -97,7 +116,12 @@ def test_dimension_attr():
+
 def test_provides():
+    '''
+    Tests for non-Fortran (pyf?) module_provides and use_provides
```
It's because I'm none too sure what the "non-Fortran" means but I'm guessing it's the "pyf" format (whatever that is).

TeranIvy replied:
[This PYF file format description link](https://datatypes.net/open-pyf-files) describes "pyf" as a F2PY Signature File, so I think the question mark can be removed. Perhaps it would be worth adding a short description of what pyf is?

arporter replied:
Good point - I've elaborated!

arporter commented on src/fparser/tests/test_api.py:
```diff
@@ -157,22 +181,38 @@ def test_provides():
-    end subroutine foo
     '''
-    tree = api.parse(source_str, isfree=True, isstrict=False, ignore_comments=False)
+    Test the walk() method of the api module
```
Done.

arporter commented on src/fparser/tests/test_api.py:
```diff
@@ -157,22 +181,38 @@ def test_provides():
+
+
+def test_caching():
+    ''' Tests relating to enabling/disabling of parser cache '''
```
Done.

arporter commented on src/fparser/api.py:
```diff
@@ -153,32 +154,35 @@ def get_reader(input, isfree=None, isstrict=None, include_dirs=None,
-    -------
-    block : `fparser.api.BeginSource`
+    :param str input: Specify a string or filename containing Fortran code.
+    :param bool isfree: Whether the Fortran source is free-format
```
Done.

arporter commented on src/fparser/api.py:
```diff
@@ -153,32 +154,35 @@ def get_reader(input, isfree=None, isstrict=None, include_dirs=None,
+    :param str input: Specify a string or filename containing Fortran code.
+    :param bool isfree: Whether the Fortran source is free-format
+    :param bool isstrict: Whether we are to strictly enforce the `isfree`
+                          setting
```
Done.

arporter commented on src/fparser/api.py:
```diff
@@ -153,32 +154,35 @@ def get_reader(input, isfree=None, isstrict=None, include_dirs=None,
+                             is required, even if the Fortran to be parsed has
+                             been seen before.
+
+    :returns: Abstract Syntax Tree of Fortran source
```
Done.

arporter commented:
All done. I've used Sphinx's autodoc to pull the interface definition of the `parse()` function into the documentation. Since the example of using that method is already in the documentation, I removed it from the docstring. Besides that I've gone through and added all the full stops :-)

Ready for 2nd review.

TeranIvy commented on doc/fparser.rst:
```diff
@@ -137,7 +137,8 @@ that reader to create the tree, e.g.:
 ::
 
-  >>> reader = FortranStringReader(code, False, True)
+  >>> reader = FortranStringReader(code)
```
Nicely done. Since you made changes in this file, could you please update the "Modified work Copyright 2017... " line at the top to 2017-2018?

arporter replied:
Well spotted - done.

arporter commented:
In fairness to @MatthewHambley, I'm going to get #70 onto master (he's already had to bring his branch up-to-date once because I took so long to review) before this one. This means I'll need to bring this branch up-to-date.

arporter commented:
#70 is now on master - I shall proceed to deal with the merge conflicts...
